serveradmin/serverdb/query_committer.py

```diff
@@ -292,43 +285,125 @@ def _upsert_attributes(attribute_lookup, changed, changed_servers):
 
 
 def _access_control(
-    entities, created_objects, changed_objects, deleted_objects
+    user, app, created_objects, changed_objects, deleted_objects
 ):
-    for server in chain(
+    """Enforce serveradmin ACLs
+
+    For servershell commits, ensure the user is allowed to make the requested
+    changes.  For adminapi commits, ensure both the user and its app are
+    allowed to make the requested changes.
+
+    Serveradmin ACLs are additive.  This means not all ACL must allow all the
+    changes a user is trying to make, but a single ACL is enough.
+
+    Note: Different ACLs may be be used to permit different parts of a commit.
+    The commit is checked per object changed in the commit.  As a result at
+    least all changes to a single object within a commit must be permissible
+    via a single ACL.
+
+    Note: when a user tries to make a change that is not permissible by any of
+    its ACLs, the error message can become rather complex, listing all the
+    reasons all the users ACLs were not applicable.
+
+    Raises PermissionDenied if a change is not permissible.
+    Returns None on success.
+    """
+
+    entities = []
+    if not user.is_superuser:
+        entities.append(
+            ('user', user, list(user.access_control_groups.all()))
+        )
+    if app and not app.superuser:
+        entities.append(
+            ('application', app, list(app.access_control_groups.all()))
+        )
+
+    # Check all objects touched by this commit
+    for obj in chain(
         created_objects.values(),
         changed_objects.values(),
         deleted_objects.values(),
     ):
+        # Check both the user and, if applicable, the users app
+        # If either doesn't have the necessary rights, abort the commit
         for entity_class, entity_name, groups in entities:
-            if not any(
-                _can_access_server(changed_objects, server, g) for g in groups
-            ):
-                raise PermissionDenied(
-                    'Insufficient access rights to server "{}" for {} "{}"'
-                    .format(server['hostname'], entity_class, entity_name)
+            acl_violations = {
+                acl: _acl_violations(changed_objects, obj, acl)
+                for acl in groups
+            }
+
+            # If all ACLs resulted in violations, none of them allowed the edit
+            # Build a verbose error message and abort the commit
+            if all(acl_violations.values()):
+                msg = (
+                    'Insufficient access rights to object "{}" for {} "{}": '
+                    .format(obj['hostname'], entity_class, entity_name)
                 )
+                for acl, violations in acl_violations.items():
+                    msg += ' '.join(violations)
```

Comment:
Maybe add newlines here to increase readability of the error messages ?

Reply:
I actually tried this initially but the django message box thingy we use collapses newlines back into spaces anyway. Further we also log this message and multi line log messages could be hard to make sense of.

Reply:
I had similar issues with errors from HWLB pushing and I went with `<pre>` tag around output. Maybe that would be an option?